src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java

```diff
@@ -0,0 +1,125 @@
+import net.neoforged.neoforge.network.payload.FrozenRegistryPayload;
+import net.neoforged.neoforge.network.payload.FrozenRegistrySyncCompletedPayload;
+import net.neoforged.neoforge.network.payload.FrozenRegistrySyncStartPayload;
+import net.neoforged.neoforge.network.payload.TierSortingRegistryPayload;
+import net.neoforged.neoforge.registries.RegistryManager;
+import net.neoforged.neoforge.registries.RegistrySnapshot;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public class ClientPayloadHandler {
+
+    private static final ClientPayloadHandler INSTANCE = new ClientPayloadHandler();
+
+    public static ClientPayloadHandler getInstance() {
+        return INSTANCE;
+    }
+
+    private final Set<ResourceLocation> toSynchronize = Sets.newConcurrentHashSet();
+    private final Map<ResourceLocation, RegistrySnapshot> synchronizedRegistries = Maps.newConcurrentMap();
+
+    private ClientPayloadHandler() {}
+
+    public void handle(FrozenRegistryPayload payload, ConfigurationPayloadContext context) {
+        synchronizedRegistries.put(payload.registryName(), payload.snapshot());
+        toSynchronize.remove(payload.registryName());
+    }
+
+    public void handle(FrozenRegistrySyncStartPayload payload, ConfigurationPayloadContext context) {
+        this.toSynchronize.addAll(payload.toAccess());
+        this.synchronizedRegistries.clear();
+    }
+
+    public void handle(FrozenRegistrySyncCompletedPayload payload, ConfigurationPayloadContext context) {
+        if (!this.toSynchronize.isEmpty()) {
+            context.packetHandler().disconnect(Component.translatable("neoforge.registries.sync.failed", this.toSynchronize.stream().map(Object::toString).collect(Collectors.joining(", "))));
+            return;
+        }
+
+        context.workHandler().submitAsync(() -> {
+            //This method normally returns missing entries, but we just accept what the server send us and ignore the rest.
+            RegistryManager.applySnapshot(synchronizedRegistries, false, false);
+
+            this.toSynchronize.clear();
+            this.synchronizedRegistries.clear();
+        }).exceptionally(e -> {
+            context.packetHandler().disconnect(Component.translatable("neoforge.registries.sync.failed", e.getMessage()));
+            return null;
+        }).thenAccept(v -> {
+            context.handler().send(new FrozenRegistrySyncCompletedPayload());
+        });
+    }
+
+    public void handle(ConfigFilePayload payload, ConfigurationPayloadContext context) {
+        ConfigSync.INSTANCE.receiveSyncedConfig(payload.contents(), payload.fileName());
+    }
+
+    public void handle(TierSortingRegistryPayload payload, ConfigurationPayloadContext context) {
+        TierSortingRegistry.handleSync(payload, context);
+    }
+
+    public void handle(AdvancedAddEntityPayload advancedAddEntityPayload, PlayPayloadContext context) {
+        context.workHandler().submitAsync(
```

Comment:
Do we want this to be execute or add some other error message like we are doing for handling registry freezing rather than this kind of just swallowing it?

Reply:
Add a disconnect on failure to process.